src/libopensc/card-fineid.c

```diff
@@ -0,0 +1,997 @@
+	int rv;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	/* Store to drv_data for later use in fineid_change_security_env() */
+	_driver_data->operation = env->operation;
+	_driver_data->key_ref_msb = env->key_ref[0];
+	_driver_data->algorithm = env->algorithm;
+	_driver_data->algorithm_flags = env->algorithm_flags;
+
+	rv = fineid_change_security_env(card);
+
+	LOG_FUNC_RETURN(card->ctx, rv);
+}
+
+
+static int
+fineid_restore_security_env(struct sc_card *card, int se_num)
+{
+	/* No need to restore after any implemented operation */
+	return SC_SUCCESS;
+}
+
+
+static int
+fineid_compute_signature(struct sc_card *card, const unsigned char *in, size_t ilen,
+		unsigned char * out, size_t olen)
+{
+	struct private_driver_data *_driver_data = (struct private_driver_data *) card->drv_data;
+	struct sc_apdu apdu;
+	unsigned char instr[SC_MAX_APDU_BUFFER_SIZE];
+	unsigned char req[SC_MAX_APDU_BUFFER_SIZE];
+	unsigned char resp[SC_MAX_APDU_BUFFER_SIZE];
+	size_t ii = 0, reqlen, orglen, blklen = 64;
+	int rv;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (!card || !in || !out) {
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+
+	else if (ilen > 96) {
+		sc_log(card->ctx, "Illegal input length %"SC_FORMAT_LEN_SIZE_T"u", ilen);
+		LOG_TEST_RET(card->ctx, SC_ERROR_INVALID_ARGUMENTS, "Illegal input length");
+	}
+
+	sc_log(card->ctx, "inlen %"SC_FORMAT_LEN_SIZE_T"u, outlen %"SC_FORMAT_LEN_SIZE_T"u", ilen, olen);
+
+	memcpy(&instr, in, ilen);
+
+	/* If no algorithm flags given, try to figure flags out from pkcs1 prefix */
+	if(fineid_get_algo(_driver_data->algorithm_flags) == FINEID_ALGO_HIGH_NA &&
+	   ilen != 20 && ilen != 28 && ilen != 32 && ilen != 48 && ilen != 64) {
+		orglen = ilen;
+		sc_log(card->ctx, "Stripping pkcs prefix, cur flags: %X, cur length: %lu",
+			_driver_data->algorithm_flags, orglen);
+
+		sc_pkcs1_strip_digest_info_prefix(&_driver_data->algorithm_flags, instr, ilen, instr, &ilen);
+		_driver_data->algorithm_flags = _driver_data->algorithm_flags | FINEID_ALGO_LOW_RSASSA_PKCS1;
+		sc_log(card->ctx, "Stripped pkcs prefix, new flags: %X, new length: %lu",
+			_driver_data->algorithm_flags, ilen);
+
+		/* If prefix was present, re-provision security env */
+		if(orglen > ilen) {
+			rv = fineid_change_security_env(card);
+			LOG_TEST_RET(card->ctx, rv, "Security env change with new algorithm failed");
+		}
+	}
+
+	/* If data does not fit in block in length */
+	if(ilen>blklen) {
+		for (ii=0; ii<ilen-blklen; ii+=blklen) {
+			/* Security operation hash
+			 *   INS 2A sec op
+			 *   P1  90 hash
+			 *   P2  80 one block of data
+			 */
+			sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0x2A, 0x90, 0x80);
+			apdu.datalen = blklen;
+			apdu.data = instr+ii;
+			apdu.lc = blklen;
+
+			sc_log(card->ctx, "Iterating at offset %lu", ii);
+			rv = sc_transmit_apdu(card, &apdu);
+			LOG_TEST_RET(card->ctx, rv, "APDU transmit failed");
+			rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+			LOG_TEST_RET(card->ctx, rv, "Block send with offset failed");
```

Comment:
is this code working? If the input is a hash, then you're hashing the hash, which will be the wrong input for the signature.